TypeHinting for public functions

huehnerhose opened the pull request:
Like mentioned in #75 I would prefer type hinting as solution. 
Since TypeHinting doesn't allow default parameter values other than null, we have to default them manually.

huehnerhose commented:
If I am not mistaken, reverting this to 12645bc should pass on php5.6 as well.

stelgenhof commented on src/Yasumi/Yasumi.php:
```diff
@@ -120,7 +120,7 @@ public static function nextWorkingDay(string $class, DateTime $startDate, int $w
 
         // Default to 1
-        $workingDays = is_null($workingDays) ? 1 : $workingDays;
+        $workingDays = $workingDays ?? 1;
```
The Null Coalescing Operator was introduced in PHP7. At the moment Yasumi still supports PHP5 so this will cause BC issues.

huehnerhose replied:
Yes. At first I didn't use null coalescing, but than I thought it would look nicer and since all my projects depend on php7.1, I did't thought twice.

huehnerhose commented:
I reverted this and switched back to the ternary version. But Travis CI doesn't seem to like this very much. I ran the test locally with success. 

Honestly I am puzzled by the errors raised, especially on the 5.6 environment claiming to expect a `Yasum\string`.

stelgenhof commented:
Your type hint uses 'string' in the current namespace 'Yasumi', which doesn't exist of course. Also 'string' as a type was only introduced as of version 7 :( See: http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration

huehnerhose commented:
Urg... didn't realized that type hinting first came for not scalar types. Sorry. 

I don't know how important php5 support ist for you / your users. If you consider it valuable we can make the check "by hand" inside the functions and manually throwing exceptions.

huehnerhose commented:
I wanted to draft a php5 version 
https://github.com/InSitu-Software/yasumi/tree/feature/typehinted5
but now all tests fail with

    Yasumi\tests\Ukraine\SecondInternationalWorkersDayTest::testSecondInternationalWorkersDay with data set #9 ('1989', DateTime Object (...))
    InvalidArgumentException: integer expected, string given

😅

stelgenhof commented:
I'll create a branch for the PHP7 development. It's something I've been planning but haven't gotten around to. Maybe better to make all your changes there. In the near future I'd like to end PHP5 support and continue to use PHP7.

stelgenhof commented:
PHP7 branch created :). Please go ahead and use that for any PHP7 related enhancements.

stelgenhof commented:
I've updated the travis configuration so it should work now.
